pkg/compactor/compactor.go

```diff
@@ -0,0 +1,263 @@
+package compactor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/thanos-io/thanos/pkg/compact"
+	"github.com/thanos-io/thanos/pkg/compact/downsample"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// Config holds the Compactor config.
+type Config struct {
+	BlockRanges          cortex_tsdb.DurationList `yaml:"block_ranges"`
+	BlockSyncConcurrency int                      `yaml:"block_sync_concurrency"`
+	ConsistencyDelay     time.Duration            `yaml:"consistency_delay"`
+	DataDir              string                   `yaml:"data_dir"`
+	CompactionInterval   time.Duration            `yaml:"compaction_interval"`
+	CompactionRetries    int                      `yaml:"compaction_retries"`
+
+	// No need to add options to customize the retry backoff,
+	// given the defaults should be fine, but allow to override
+	// it in tests.
+	retryMinBackoff time.Duration `yaml:"-"`
+	retryMaxBackoff time.Duration `yaml:"-"`
+}
+
+// RegisterFlags registers the Compactor flags.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.BlockRanges = cortex_tsdb.DurationList{2 * time.Hour, 12 * time.Hour, 24 * time.Hour}
+	cfg.retryMinBackoff = 10 * time.Second
+	cfg.retryMaxBackoff = time.Minute
+
+	f.Var(&cfg.BlockRanges, "compactor.block-ranges", "Comma separated list of compaction ranges expressed in the time duration format")
```

Comment:
Could you add the default values to the help text?

Reply:
There actually no need. When the `--help` is generated, defaults are displayed reading them from the values set:

```
 -compactor.block-ranges value
        Comma separated list of compaction ranges expressed in the time duration format (default 2h0m0s,12h0m0s,24h0m0s)
```

Reply:
Awesome